Expose the GetShortestBitLength, GetByteCount, and WriteLittleEndian APIs

tannergooding opened the pull request:
Still pending are the equivalent APIs for getting the exponent/significand for floating-point numbers and the Create* API changes, will be in follow up PRs to help keep size down

tannergooding committed:
Expose GetShortestBitLength, GetByteCount, and WriteLittleEndian APIs

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Char.cs:
```diff
@@ -1671,6 +1675,27 @@ static bool INumber<char>.TryParse(ReadOnlySpan<char> s, NumberStyles style, IFo
+        /// <inheritdoc cref="INumber{TSelf}.TryWriteLittleEndian(Span{byte}, out int)" />
+        bool INumber<char>.TryWriteLittleEndian(Span<byte> destination, out int bytesWritten)
+        {
+            if (destination.Length >= sizeof(char))
+            {
+                ushort value = BitConverter.IsLittleEndian ? m_value : BinaryPrimitives.ReverseEndianness(m_value);
+                Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(destination), value);
+
+                bytesWritten = sizeof(char);
+                return true;
+            }
+            else
+            {
+                bytesWritten = 0;
+                return false;
+            }
```
Did you consider using BinaryPrimitives for these implementations?  I'm assuming you chose not to because those TryWriteXxLittleEndian methods don't take an out bytesWritten and thus you'd need to implement that manually anyway?

tannergooding replied:
Right, it would've done duplicate work/checks + not cover everything I need to do here. So it would ultimately not been any smaller or faster.

tannergooding commented:
CC. @lewing @BrzVlad

This is triggering a mono-runtime assert for wasm and impacts the generic math API surface we want to preview at build.

dakersnar commented on src/libraries/System.Runtime/tests/System/ByteTests.GenericMath.cs:
```diff
@@ -116,6 +115,16 @@ public static void TrailingZeroCountTest()
+        [Fact]
+        public static void GetShortestBitLengthTest()
+        {
+            Assert.Equal(0x00, BinaryIntegerHelper<byte>.GetShortestBitLength((byte)0x00));
```
I was originally confused by your use of hex on the left-hand side, but I can see why it helps with manually calculating the expected result (counting groups of 16 bits).

dakersnar reviewed (approved):
Looks good!

tannergooding commented:
Worked around some test failures (and logged corresponding issues). Also fixed an error where these APIs shouldn't be on float/double/decimal/Half. The floating-point types are getting separate APIs to get the exponent and significand instead, so that the value can be understood.

tannergooding commented:
No changes, just squashed commits together and rebased ontop of https://github.com/dotnet/runtime/pull/68043 now that it's merged.

This is now "just" the `GetShortestBitLength`, `GetByteCount`, and `WriteLittleEndian` APIs + relevant tests and doesn't include the WASM disablement since https://github.com/dotnet/runtime/issues/67999 repros even with them disabled

tannergooding committed:
Merge remote-tracking branch 'dotnet/main' into generic-math

tannergooding commented:
Merge with main to pickup https://github.com/dotnet/runtime/pull/68058
